compiler/rustc_mir_transform/src/check_enums.rs

```diff
@@ -0,0 +1,583 @@
+    tcx: TyCtxt<'tcx>,
+    local_decls: &'a mut LocalDecls<'tcx>,
+    typing_env: TypingEnv<'tcx>,
+    enums: Vec<EnumCheckType<'tcx>>,
+}
+
+impl<'a, 'tcx> EnumFinder<'a, 'tcx> {
+    fn new(
+        tcx: TyCtxt<'tcx>,
+        local_decls: &'a mut LocalDecls<'tcx>,
+        typing_env: TypingEnv<'tcx>,
+    ) -> Self {
+        EnumFinder { tcx, local_decls, typing_env, enums: Vec::new() }
+    }
+
+    /// Returns the found enum creations and which checks should be inserted.
+    fn into_found_enums(self) -> Vec<EnumCheckType<'tcx>> {
+        self.enums
+    }
+
+    /// In some cases the enum discriminant is stored in a tag that is represented by
+    /// primitive. This method returns the actual discriminant type and size for that
+    /// tag.
+    fn tag_type_and_size_for_primitive(&self, primitive: Primitive) -> TyAndSize<'tcx> {
+        let discr_ty = match primitive {
+            Primitive::Int(Integer::I8, false) => self.tcx.types.u8,
+            Primitive::Int(Integer::I8, true) => self.tcx.types.i8,
+            Primitive::Int(Integer::I16, false) => self.tcx.types.u16,
+            Primitive::Int(Integer::I16, true) => self.tcx.types.i16,
+            Primitive::Int(Integer::I32, false) => self.tcx.types.u32,
+            Primitive::Int(Integer::I32, true) => self.tcx.types.i32,
+            Primitive::Int(Integer::I64, false) => self.tcx.types.u64,
+            Primitive::Int(Integer::I64, true) => self.tcx.types.i64,
+            Primitive::Pointer(AddressSpace(0)) => self.tcx.types.usize,
+            _ => bug!("Invalid discriminant {:#?}", primitive),
+        };
+        let Ok(discr_layout) = self.tcx.layout_of(self.typing_env.as_query_input(discr_ty)) else {
+            bug!("Invalid discriminant {:#?}", primitive);
+        };
+        TyAndSize { ty: discr_ty, size: discr_layout.size }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for EnumFinder<'a, 'tcx> {
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        if let Rvalue::Cast(CastKind::Transmute, op, ty) = rvalue {
+            let ty::Adt(adt_def, _) = ty.kind() else {
+                return;
+            };
+            if !adt_def.is_enum() {
+                return;
+            }
+
+            let Ok(enum_layout) = self.tcx.layout_of(self.typing_env.as_query_input(*ty)) else {
+                return;
+            };
+            let Ok(op_layout) = self
+                .tcx
+                .layout_of(self.typing_env.as_query_input(op.ty(self.local_decls, self.tcx)))
+            else {
+                return;
+            };
+
+            match enum_layout.variants {
+                Variants::Empty if op_layout.is_uninhabited() => return,
+                // An empty enum that tries to be constructed from an inhabited value, this
+                // is never correct.
+                Variants::Empty => {
+                    // The enum layout is uninhabited but we construct it from sth inhabited.
```

Comment:
Isn't this case already detected statically?

Reply:
Yes it is. I was wondering whether I should put it here as it already creates a warning. But I thought that it would be good to have it for the sake of completeness. I am happy to remove it if you think that's better.

Reply:
Oh my it only gets a warning. Then I agree it's better to have the actual runtime check.